Add baseline correct widget

atravert opened the pull request:
- [X] Tests added.  Note:  I dont't know how to trigger methods calld by on_click() 
- [X] The docstrings have been tested with the script `script/validate_docstrings.py`: note that there is no real example as it works only in notebook not on python line command. 
- [X] User-visible changes have been documented in `CHANGELOG.md` 
- [X] The new methods have been listed in `docs/userguide/reference/api.rst`.

fernandezc commented:
> * [x]  Tests added.  Note:  I dont't know how to trigger methods calld by on_click()

I think it is probably necessary to mock on_click() function call.  
For example look at line 91 in test_importer.py, where  I am mocking the opening of a dialog box. Something similar could be used, I presume.

fernandezc commented:
> > * [x]   Tests added.  Note:  I dont't know how to trigger methods calld by on_click()
> 
> I think it is probably necessary to mock on_click() function call. For example look at line 91 in test_importer.py, where I am mocking the opening of a dialog box. Something similar could be used, I presume.

And line 253 for instance (monkeypatch)

fernandezc commented:
> > > * [x]    Tests added.  Note:  I dont't know how to trigger methods calld by on_click()
> > 
> > 
> > I think it is probably necessary to mock on_click() function call. For example look at line 91 in test_importer.py, where I am mocking the opening of a dialog box. Something similar could be used, I presume.
> 
> And line 253 for instance (monkeypatch)

Actually this use a pytest fixture 
test_importer(monkeypatch)

fernandezc commented on CHANGELOG.md:
```diff
@@ -1,5 +1,10 @@
 # What's new
 
+## version 0.3.3
```
Comme je te le mentionnais ici https://github.com/spectrochempy/spectrochempy/pull/387#discussion_r798771801
Ce changhe log n'est pas correct. Lors de ton dernier merge la version 0.3.2 existait déjà! 
Donc le change log devrait etre: 
```md 
## VERSION 0.3

### version 0.3.3

#### NEW FEATURES

* add `BaselineCorrector()` widget

* add `download_nist_ir()` to download IR spectra from NIST/webbook

* `read_spa()` now allows extracting sample and background interferograms

#### BUGS FIXED

* fix gettingstarted/overview.py after IRIS refactoring

### version 0.3.2

#### NEW FEATURES

* Add a log file (rotating)

#### BUGS FIXED

* TQDM progress bar
 ...
```

fernandezc replied:
Pour le reste j'essayerai d'avancer dans le week end...

fernandezc replied:
D'ailleurs, tu as vu que j'ai changé un peu le format dans le post ci dessus.
En fait je trouve que c'est mieux comme ça: 

Voici le changelog complet que j'aimerais utiliser dorénavant. Si ca te convient tu peut le mettre comme ca dans ce pPR.

```md
---
## VERSION 0.3

### version 0.3.3

#### NEW FEATURES

* add `BaselineCorrector()` widget

* add `download_nist_ir()` to download IR spectra from NIST/webbook

* `read_spa()` now allows extracting sample and background interferograms

#### BUGS FIXED

* fix gettingstarted/overview.py after IRIS refactoring

### version 0.3.2

#### NEW FEATURES

* Add a log file (rotating)

#### BUGS FIXED

* TQDM progress bar

* Fix #360 : write a dataset with a specified filename do not open a dialog except if the file already exists and if
  confirm=True is set.

* `read` and `read_dir` now ignore non readable files.

* `read_labspec` now ignore non-labspec .txt files.

* Fix #296 : IRIS and quadprog version.

* Fix #375 : plotting issues.

### Version 0.3.1

#### NEW FEATURES

* Added a `show_versions` method in the API.

* Improved bug reports and pull request templates. Requests for
  help are now made in github discussions.

* Docs API reference has been hopefully improved.

### Version 0.3.0

#### NEW FEATURES

* Package refactoring which may break previous  behaviour. This is why we
  updated the minor version number from 0.2 to 0.3.

* GRAMS/Thermo Galactic .spc file reader.

* Fitting models updated and tested.

#### BUGS FIXED

* Bug in check_updates preventing working without connection.

---
## VERSION 0.2

### Version 0.2.23

#### BUGS FIXED

* Workflow/Codeclimate issues

### Version 0.2.22

#### BUGS FIXED

* QT save_dialog.

* Plot_multiple bug.

### Version 0.2.21

#### NEW FEATURES

* Indexing or slicing a NDArray with quantities is now possible.

* MatPlotLib Axes are subclassed in order to accept quantities for method arguments.

#### BUGS FIXED

* NDArray constructor now accept a homogeneous list of quantities as data input. Units are set accordingly.

* Qt Dialogs. This is related to issue #198, as tk dialogs can be replaced by Qt when working with a terminal.

* Custom exception.

* Qt Dialogs. This is related to issue #198, as tk dialogs can be replaced by Qt when working with a terminal.

* Doc display problems.

### Version 0.2.18

#### NEW FEATURES

* pip installation now possible

* Some code revision

* NNMF revision

* Documentation improvement

#### BUGS FIXED

* Issue #310

* The order of multicoordinates for a single dimension

* Integrate methods to avoid code-climate warnings (duplicate code)

* Documentation for the integrate methods

* skipping test_sh under windows


### Version 0.2.17

#### NEW FEATURES

* OPUS file reader: add filenames as labels.

* OMNIC file reader: Documented more .spa header keys.

#### BUGS FIXED

* Compatibility with matplotlib 3.5 (issue #316).

* Datasets were not properly centered in PCA analysis.

* Comparing dataset with only labels coordinates was failing.

* Issue #322: mean and other API reduce methods were sometimes failing.

### Version 0.2.16

#### NEW FEATURES

* IRIS: Added 1D datasets.

* IRIS: Added kernel function for diffusion .

* EFA: Added indication of progress.

* Cantera: Added differential evolution algorithm in cantera utilities.

* Cantera: Added PFR object in cantera utilities.

* DOC: Added list of papers citing spectrochempy.

* Github action workflows to test, build and publish conda package and docs in replacement of Travis CI.

* Use CodeClimate to show Coverage info

#### BUGS FIXED

* IRIS example after modification of readers.

* IRIS: automatic search of the L-curve corner.

* MCR-ALS returns the 'soft' concentration matrix.

* Document building configuration after update of external packages.

* DOC: several broken links.

* Baseline correction default changed.

* Compatibility with newest change in Colab

### Version 0.2.15

#### NEW FEATURES

* Added a baseline correction method: `basc`.

* Baseline ranges can be stored in meta.regions['baseline'] - basc will recognize them.

#### BUGS FIXED

* Comparison of dataset when containing metadata in testing functions.

* Project.

* Bug in the `to` function.

### Version 0.2.14

#### NEW FEATURES

* A default coordinate can now be selected for multiple coordinates dimensions.

#### BUGS FIXED

* Alignment along several dimensions (issue #248)

* to() and ito() methods to work correctly (issue #255)

* Baseline correction works on all dimensions

### Version 0.2.13

#### BUGS FIXED

* Solved the problem that reading of experimental datasets was too slow in v.0.2.12.

### Version 0.2.12

#### BUGS FIXED

* LinearCoord operations now working.

* Baseline default now "sequential" as expected.

**WARNING**: It was wrongly set to "mutivariate" in previous releases, so you should  expect some difference with processing you may have done before.

* Comparison of coordinates now correct for mathematical operations.

* Alignment methods now working (except for multidimensional alignment).

### Version 0.2.11

#### BUGS FIXED

* Plot2D now works when more than one coord in 'y' axis (#238).

* Spectrochempy_data location has been corrected (#239).

### Version 0.2.10

#### NEW FEATURES

* All data for tests and examples are now external.

They are now located in a separate conda package: `spectrochempy_data`.

* Installation in Colab with Examples is now supported.

#### BUGS FIXED

* Read_quadera() and examples now based on a correct asc file

### Version 0.2.9

#### BUGS FIXED

* Hotfix regarding display of NMR x scale

### Version 0.2.8

#### NEW FEATURES

* Added write_csv() dir 1D datasets

* Added read_quadera() for Pfeiffer Vacuum's QUADERA® MS files

* Added test for trapz(), simps(), readquadera()

* Improved displaying of Interferograms

#### BUGS FIXED

* Problem with trapz(), simps()

* interferogram x scaling

### Version 0.2.7

#### NEW FEATURES

* Test and data for read_carroucell(), read_srs(), read_dso()

* Added NMR processing of 2D spectra.

* Added FTIR interferogram processing.

#### BUGS FIXED

* Problem with read_carroucell(), read_srs(), read_dso()

* Colaboratory compatibility

* Improved check updates

### Version 0.2.6

#### NEW FEATURES

* Check for new version on anaconda cloud spectrocat channel.

* 1D NMR processing with the addition of several new methods.

* Improved handling of Linear coordinates.

#### BUGS FIXED

* Adding quantity to datasets with different scaling (#199).

* Math now operates on linear coordinates.

* Compatibility with python 3.6

### Version 0.2.5

#### NEW FEATURES

* Docker image building.

* Instructions to use it added in the documentation.

* Cantera installation optional.

* Use of pyqt for matplotlib optional.

#### BUGS FIXED

* Added fonts in order to solve missing fonts problems on Linux and windows.

### Version 0.2.4

#### NEW FEATURES

* Documentation largely revisited and hopefully improved. *Still some work to be done*.

* NDMath (mathematical and dataset creation routines) module revisited. *Still some work to be done*.

* Changed CoordRange behavior.

#### BUGS FIXED

* Problem with importing the API.

* Dim handling in processing functions.

### Version 0.2.0

#### NEW FEATURES

* Copyright update.

* Requirements and env yml files updated.

* Use of the coordinates in math operation improved.

* Added ROI and Offset properties to NDArrays.

* Readers / Writers revisited.

* Bruker TOPSPIN reader.

* Added LabSpec reader for .txt exported files.

* Simplified the format of scp file - now zipped JSON files.

* Rewriting json serialiser.

* Add function pathclean to the API.

* Add some array creation function to NDMath.

* Refactoring plotting preference system.

* Baseline correction now accepts single value for ranges.

* Add a waterfall plot.

* Refactoring plot2D and 1D methods.

* Added Simpson'rule integration.

* Addition of multiple coordinates to a dimension works better.

* Added Linear coordinates (EXPERIMENTAL).

* Test for NDDataset dtype change at initialization.

* Added subdir of txt files in ramandata.

* Comparison of datasets improved in testing.py.

* Comparison of datasets and projects.

#### BUGS FIXED

* Dtype parameter was not taken into account during initialization of NDArrays.

* Math function behavior for coords.

* Color normalization on the full range for colorscale.

* Configuration settings in the main application.

* Compatibility read_zip with py3.7.

* NDpanel temporary removed from the master.

* 2D IRIS.

* Trapz integration to return NDDataset.

* Suppressed a forgotten sleep statement that was slowing down the SpectroChemPy initialization.

* Error in SIMPLISMA (changed affectations such as C.data[...] = something by C[...] = something.

* Cleaning mplstyle about non-style parameters and corrected makestyle.

* Argument of set_xscale.

* Use read_topspin instead of the deprecated function read_bruker_nmr.

* Some issues with interactive baseline.

* Baseline and fitting tutorials.

* Removed dependency of isotopes.py to pandas.

---
## VERSION 0.1

### Version 0.1.x

* Initial development versions.
```

fernandezc commented:
Arnaud, qu'en est-il de la performance de ce widget... l'affichage n'est pas trop lent? Je vais essayer bien sûr mais peut etre peut tu commenter déjà la dessus. Si c'est trop lent, on pourrait essayer d'autre librairies comme poorly ou bokeh pour cette tache particulière. J'avais commencer à preparer l'utilisation de Plotlly mais pas beaucoup avancé pour le moment (en fait un peu mis de coté depuis que je n'ai pas travaillé sur l'application dash)

fernandezc commented:
> comme poorly ou bokeh 
plotly bien sûr (foutue correction automatique!)

atravert commented:
J'ai essayé sur une dizaine de spectres et ça me parait bien. Au delà je 
ne sais pas, mais de toute façon l'idée ici est que l'utilisateur 
utilise a bon escient le slicing (si il a 1000 spectres, il entre [0: 
-1: 100] dans y slice ce qui lui donne déjà un echantillon  
représentatif à mon avis. De toute façon au delà, l'examen visuel de 
plusieurs dizaines de spectres + leur ligne de base n'est pas très 
réaliste non plus -- toujours à mon avis . Ensuite, il ajuste ensuite 
ses paramètres (range, method, order, ....) par essais erreur: la 
correction de ligne de base et l'affichage ne portent que sur ses 10 
spectres. Une fois qu'il est content, il fait un dernier process en 
changeant uniquement le step du y slice:  [0: -1: 1].


Le 25/02/2022 à 15:33, Christian Fernandez a écrit :
>
> Arnaud, qu'en est-il de la performance de ce widget... l'affichage 
> n'est pas trop lent? Je vais essayer bien sûr mais peut etre peut tu 
> commenter déjà la dessus. Si c'est trop lent, on pourrait essayer 
> d'autre librairies comme poorly ou bokeh pour cette tache 
> particulière. J'avais commencer à preparer l'utilisation de Plotlly 
> mais pas beaucoup avancé pour le moment (en fait un peu mis de coté 
> depuis que je n'ai pas travaillé sur l'application dash)
>
> —
> Reply to this email directly, view it on GitHub 
> <https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fspectrochempy%2Fspectrochempy%2Fpull%2F397%23issuecomment-1050905202&data=04%7C01%7Carnaud.travert%40ensicaen.fr%7C1d604554fad94f05816108d9f86bd74a%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C637813964387375379%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=U%2BXKGVfiIraEjCch0qc8I%2BTMJ57EkVdtmDvR%2BxrRK9U%3D&reserved=0>, 
> or unsubscribe 
> <https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FACNBNCTLBVXO4GQMHGLHHBDU46HM3ANCNFSM5PKE7BBA&data=04%7C01%7Carnaud.travert%40ensicaen.fr%7C1d604554fad94f05816108d9f86bd74a%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C637813964387375379%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=TwB0WtBk12IVONN3OFgrDGWWUq%2B51tfojkupr%2FyElT0%3D&reserved=0>.
> Triage notifications on the go with GitHub Mobile for iOS 
> <https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fapps.apple.com%2Fapp%2Fapple-store%2Fid1477376905%3Fct%3Dnotification-email%26mt%3D8%26pt%3D524675&data=04%7C01%7Carnaud.travert%40ensicaen.fr%7C1d604554fad94f05816108d9f86bd74a%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C637813964387375379%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=ZEIVWEBn7rDWYsWJbSWCtCQmGvQjz%2BypCazi%2BXk3lvE%3D&reserved=0> 
> or Android 
> <https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fplay.google.com%2Fstore%2Fapps%2Fdetails%3Fid%3Dcom.github.android%26referrer%3Dutm_campaign%253Dnotification-email%2526utm_medium%253Demail%2526utm_source%253Dgithub&data=04%7C01%7Carnaud.travert%40ensicaen.fr%7C1d604554fad94f05816108d9f86bd74a%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C637813964387375379%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=Rs1927mgrv%2FynrzdVrhoNZcwcm28M%2Fu6n9bQN0CbFvs%3D&reserved=0>. 
>
> You are receiving this because you authored the thread.Message ID: 
> ***@***.***>
>

atravert commented:
In jupyter notebook you must use the widget backend: 

℅matplotlib widget

In princiole the plits should be updated

In jupyter lab I had the same behaviour as yours.

Le 26 février 2022 17:02:47 GMT+01:00, Christian Fernandez ***@***.***> a écrit :
>I have checkout this PR and tried baseline.py in the user guide, in
>Jupyter Lab and in classical notebook.
>In classical notebook, the display is not updated, instead, new plots
>are done each time I process (I assume this is not what is expected).
>In Jupyter lab, there is no new display after the initial one (but I
>can see them in logs... not sure this is due to a particular
>configuration on my side!).
>
>—
>Reply to this email directly, view it on
>GitHub<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fspectrochempy%2Fspectrochempy%2Fpull%2F397%23issuecomment-1052245478&data=04%7C01%7Carnaud.travert%40ensicaen.fr%7C5b9c0b31702e468e562608d9f9416f6e%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C637814881736156907%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=GzOeeEjTxZG9z4bJO7ICsC3htDpOkOB%2BxAQU4QjkF04%3D&reserved=0>,
>or
>unsubscribe<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FACNBNCSMC3O63JQ4R6U5Q6TU5D2SPANCNFSM5PKE7BBA&data=04%7C01%7Carnaud.travert%40ensicaen.fr%7C5b9c0b31702e468e562608d9f9416f6e%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C637814881736156907%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=esGxuyOSUqMq5ryCn0mOrN1crv7t2%2FMnr%2FetY1UaSiE%3D&reserved=0>.
>Triage notifications on the go with GitHub Mobile for
>iOS<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fapps.apple.com%2Fapp%2Fapple-store%2Fid1477376905%3Fct%3Dnotification-email%26mt%3D8%26pt%3D524675&data=04%7C01%7Carnaud.travert%40ensicaen.fr%7C5b9c0b31702e468e562608d9f9416f6e%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C637814881736156907%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=Fo36G%2F3L75N6Atnr44o80Hvod9kRkf9PW7Kuk1ODU7Y%3D&reserved=0>
>or
>Android<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fplay.google.com%2Fstore%2Fapps%2Fdetails%3Fid%3Dcom.github.android%26referrer%3Dutm_campaign%253Dnotification-email%2526utm_medium%253Demail%2526utm_source%253Dgithub&data=04%7C01%7Carnaud.travert%40ensicaen.fr%7C5b9c0b31702e468e562608d9f9416f6e%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C637814881736156907%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=mvtuJB%2FzaY3DvgD3yr3SsrkWIvl%2Bi2U5zvS1DRVQbyo%3D&reserved=0>.
>You are receiving this because you authored the thread.Message ID:
***@***.***>

-- Envoyé depuis /e/ Mail.

fernandezc commented:
Yes, I tried both with and without the widget backend.. For me no difference.
I remember that at the beginning I was using this backend by default when initialising the API, but I had problems (not really remember which one), so I decided to simplify and put inline by default. 
But no hurry, we will see that next week.

atravert commented:
I have tried with various configs:
- python 3.7 to python 3.9 
- ipywidgets: 7.5.1 to 7.6.5  
- ipympl 0.5.8 to 0.8.8
- jupyterlab: 3.30

and always got correct behaviour in jupyter notebook  (not jupyter lab for which it does not work properly). The doc now clearly states that it should be run on jupyter notebook not jupyter lab. 

FYI my current configuration is:


Does it seem OK for you ?

fernandezc commented:
> Does it seem OK for you ?

For me, as told before, it is not working (on Mac) as expected even in notebooks. Could you try with Linux?

atravert commented:
edit: sorry it was the old configutration  (which woke). The oine with update packages (which still wok) is:

```
INSTALLED VERSIONS
------------------
commit: None
python: 3.9.10 | packaged by conda-forge | (main, Feb  1 2022, 21:22:07) [MSC v.1929 64 bit (AMD64)]
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 142 Stepping 12, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: ('en_US', 'ISO8859-1')
conda-forge: None
spectrocat: None
defaults: None
quadprog: None
brukeropusreader: 1.3
quaternion: 2021.11.4.15.26.3
colorama: 0.4.4
dill: 0.3.4
ipython: None
jinja2: 3.0.3
matplotlib: 3.5.1
numba: 0.55.1
numpy: 1.21.5
orderedset: 2.0.3
pint: 0.16.1
plotly: 4.14.3
requests: 2.24.0
scipy: 1.7.1
tqdm: 4.51.0
traitlets: 5.0.5
traittypes: 0.2.1
xlrd: 1.2.0
spectrochempy_data: None
setuptools: 60.9.3
setuptools_scm: 4.1.2
git: 3.1.26
ipywidgets: 7.6.5
ipympl: 0.5.8
jupyterlab: 3.3.0
nodejs: 0.1.1
pip: 21.3.1
pip:: None
git+https://github.com/jjhelmus/nmrglue.git: None
spectrochempy: 0.2.16.dev29g
```

atravert commented:
> For me, as told before, it is not working (on Mac) as expected even in notebooks. Could you try with Linux?

unfortunately I have no linux here...  Let's try the following: I merge the PR on the dev versioçn and ask Olivier (who works on mac) to test it, and philippe on win to test it on their side  ?

fernandezc commented:
> > For me, as told before, it is not working (on Mac) as expected even in notebooks. Could you try with Linux?
> 
> unfortunately I have no linux here... Let's try the following: I merge the PR on the dev versioçn and ask Olivier (who works on mac) to test it, and philippe on win to test it on their side ?

I can try tonight on linux and windows on my side (virtual systems). If it works we can merge. OK?

atravert commented:
Yes, fine :-)

fernandezc commented:
I have tried with Windows 10 on parallel desktop: 

INSTALLED VERSIONS

commit: 484886bcb770c260f1464a5b90c04bcb7bf86dda
python: 3.9.0 | packaged by conda-forge | (default, Nov 26 2020, 07:53:15) [MSC v.1916 64 bit (AMD64)]
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 158 Stepping 13, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: ('en_US', 'ISO8859-1')

conda-forge: None
spectrocat: None
defaults: None
quadprog: 0.1.8
brukeropusreader: 1.3
quaternion: 2020.9.5.14.42.2
colorama: 0.4.4
dill: 0.3.3
ipython: None
jinja2: 2.11.2
matplotlib: 3.5.1
numba: 0.55.1
numpy: 1.21.5
orderedset: 2.0.3
pint: 0.16.1
plotly: 4.14.3
requests: 2.25.1
scipy: 1.7.3
tqdm: 4.56.0
traitlets: 5.0.5
traittypes: 0.2.1
xlrd: 2.0.1
spectrochempy_data: None
setuptools: 52.0.0.post20210125
setuptools_scm: 5.0.1
git: None
ipywidgets: 7.6.3
ipympl: 0.6.3
jupyterlab: 3.0.5
nodejs: None
pip: 21.0
pip:: None
git+https://github.com/jjhelmus/nmrglue.git: None
spectrochempy: (Can't determine version string)

----

Unfortunately I got the same behavior. When I click on `process`  new figure are added - no update of the existing one.
May be it is specific to MAC, I don't know.  I did not try with linux yet (I'll will do it tomorrow)

Additionaly when I change x slice and hit process I get this error:
<img width="1191" alt="image" src="https://user-images.githubusercontent.com/2428474/156632098-622bf5fc-c73e-4f11-89f8-b6bafedf2f6b.png">




So, it looks like the PR is not yet ready for merging.  I can try to work on it this week-end probably. may be we need to add some tests to check the behavior on different system.

fernandezc commented:
There is also an error when you click on `save_as`, and then `cancel`in the dialog box.
```
AttributeError                            Traceback (most recent call last)
File ~/Dropbox/SCP/spectrochempy/spectrochempy/widgets/baselinecorrector.py:254, in BaselineCorrector.save_clicked(self, b)
    253 def save_clicked(self, b):
--> 254     self.corrected.write()

File ~/Dropbox/SCP/spectrochempy/spectrochempy/core/writers/exporter.py:142, in write(dataset, filename, **kwargs)
     98 """
     99 Write  the current dataset.
    100 
   (...)
    139 'opus.scp'
    140 """
    141 exporter = Exporter()
--> 142 return exporter(dataset, filename, **kwargs)

File ~/Dropbox/SCP/spectrochempy/spectrochempy/core/writers/exporter.py:63, in Exporter.__call__(self, *args, **kwargs)
     60     return filename
     62 except Exception as e:
---> 63     raise e

File ~/Dropbox/SCP/spectrochempy/spectrochempy/core/writers/exporter.py:57, in Exporter.__call__(self, *args, **kwargs)
     55 if kwargs.get("suffix", ""):
     56     filename = filename.with_suffix(kwargs.get("suffix", ""))
---> 57 protocol = self.protocols[filename.suffix]
     58 write_ = getattr(self, f"_write_{protocol}")
     59 write_(self.object, filename, **kwargs)

AttributeError: 'NoneType' object has no attribute 'suffix'
```

atravert commented:
OK. I have fixed 2 errors with the last commit
- The error related to pressing cancel is fixed
- The other one:
![image](https://user-images.githubusercontent.com/10098314/156655500-0ba420d3-f36c-4c82-9736-7896e9bf80e4.png)

is also fixed (when the slicing returns None because indexes have been untintentionally used instead of floats, e.g. (3000:900;1)  instead of [3000.0:900.0:1],  the warning message is still issued, this is an expected behaviour)

- for the plot update at each re-processing, I just don't now how to reproduce this error as it works fine on my computer :-).  
The expected behavu=iour is that once processed has been done once and went ok, `self._done` is set to `True` (line 203). Then when re-processing, `self.blcorrect_and_plot(clear=True)` (line 246) is called, and the output should be cleared before multiplot() is called (lines 185-188): 
```
          with self._output:
                if clear:
                    clear_output(True)
                axes = multiplot(
                ..etc...
```
I suppose that the problem comes from `clear_output(True)` which does not work as it should... the ipywdget doc is here: https://ipywidgets.readthedocs.io/en/latest/examples/Output%20Widget.html

atravert commented:
PS: in the current PR it is the  IPython.display.clear_output() which is called, but using Output widget method instead: 

```
            with self._output:
                if clear:
                    self._output.clear_output()
                axes = multiplot(
                    ....
```

also works. Maybe this can solve the jssue ?

fernandezc commented:
> PS: in the current PR it is the IPython.display.clear_output() which is called, but using Output widget method instead:
> 
> ```
>             with self._output:
>                 if clear:
>                     self._output.clear_output()
>                 axes = multiplot(
>                     ....
> ```
> 
> also works. Maybe this can solve the jssue ?

No I tried.  Same result!

atravert commented:
OK, thank you very much for your time !, let me know if I can help.

fernandezc commented:
I let you try with Jupiter lab, and modify docs if necessary.
%matplotlib widget is of course necessary in both notebook and lab case. 
%matplotlib notebook also works but only for notebooks.

atravert commented:
now the plots do not appear anymore, neither in jupyer notebook, nor jupyter lab, and whatever the backend. For inline, it plots only the sresulf of the first 'process'. I will look on my config this week end :-)

fernandezc commented:
> now the plots do not appear anymore, neither in jupyer notebook, nor jupyter lab, and whatever the backend. For inline, it plots only the result of the first 'process'. I will look on my config this week end :-)

May be remove the clear_output : If think it is not necessary anymore. 
Anyway, very strange issues!

fernandezc commented:
> > now the plots do not appear anymore, neither in jupyer notebook, nor jupyter lab, and whatever the backend. For inline, it plots only the result of the first 'process'. I will look on my config this week end :-)
> 
> May be remove the clear_output : If think it is not necessary anymore. Anyway, very strange issues!

Sorry it is needed!

fernandezc commented:
May be you can try again. I suppressed the need of clear output, which I think explain the issue in your side.

fernandezc commented:
I did some additional change regarding the slicing to keep compatibility of indexes x slicing with numpy. [0:100] was returning only 99 values. Now it returns 100.

fernandezc commented:
additionally location slicing is now also possible for y (even if it is probably no so useful)

fernandezc commented:
> * 
> * The other one:
>   ![image](https://user-images.githubusercontent.com/10098314/156655500-0ba420d3-f36c-4c82-9736-7896e9bf80e4.png)
> 
> is also fixed (when the slicing returns None because indexes have been untintentionally used instead of floats, e.g. (3000:900;1) instead of [3000.0:900.0:1], the warning message is still issued, this is an expected behaviour)

For this problem, yes it is correct in the case you mentioned, but it happens also when after the first display, you try to reduce the size of x window for example from [5999.56 : 649.9 : 1] to [5000.56 : 649.9 : 1]. This is because the range (10% at the left side is no more in the window). I will try to solve this problem.

atravert commented:
I managed to got it working finally... apparently it was related to a particular combination of versions of ipympl / jupyterlab /  ... which worked fine before the changes you've made and not anymore !

now, if you want,  I can work on the remaining issues: 
- error when processing with slice outside the range; -- here it is an error that we would expect from underlying baseline correction:   
```
A = scp.read("irdata/nh4y-activation.SPG")
blc = scp.BaselineCorrection(A[:, 3000.0:1000.])
ranges = ([4000., 3600.0], [1200., 1100.])
blc.compute(*ranges)
```
returns a similar error message. So we should certainly fix this in BaselineCorrection.compute(). In such a situation  there are many possibilities:
     -  raise an error with a specific message  ("a range is outside the dataset limit)
     - return None and raise awarning with a specific message  ("a range is outside the dataset limit)
     - ignore ranges outside the limits and:
     - if a range is still valid:  make baseline correction from this range and raise a warning
      - if all range outside raise an error 
      - return None and raise a worning
      - ....

I think that the 1st or 2nd solution are the best: it is the user responsibility to check his input, and there is no ambiguity. But maybe you have another opinion on this ?

- Returning to the widget;  a second issue is to clear also previous warning/error messages when 're-processing'. I can also look at this. Let me know !

fernandezc commented:
Yes, I agree in the example you give, because the user sets the wrong range, that is the reason for the warning. But in the case I cited, it is not an intentional error by the user since the initial range is set to 10% on each side by default, if I understood correctly. 
 The following test shows the problem when I simulate changing the x-split:
```
def test_range_issue():
    X = scp.read_omnic("irdata/nh4y-activation.spg")
    out = scp.BaselineCorrector(X) 
    out._x_limits_control.value = "[5000.56 : 649.9 : 1] "
    out.process_clicked()  
# this generates a warning (it should not and the range should be corrected to a point at the window boundary, for example)
```

of course, if I do this
```
def test_range_issue() :
initial_range = (
        [5900.0, 5400.0],
        [4000.0, 4500.0],
        4550.0,
        [2100.0, 2000.0],
        [1550.0, 1555.0],
        [1250.0, 1300.0],
        [800.0, 850.0],
        )
    X = scp.read_omnic("irdata/nh4y-activation.spg")
    out = scp.BaselineCorrector(X, initial_ranges=initial_ranges)
    out._x_limits_control.value = "[5000.56 : 649.9 : 1] "
    out.process_clicked()  
# now the warning is issued but I think it is the user's responsibility to change the range values before starting.
```

fernandezc commented:
Yes of course you can work on this. This was your idea. I just wanted to help

fernandezc commented:
> * Returning to the widget;  a second issue is to clear also previous warning/error messages when 're-processing'. I can also look at this. Let me know !

For this I think it is because I do not use the clear_output. But it can be put back if it works for you (I thought the problem you had could be related)

fernandezc commented:
> What I can do in my side it to work on the tests, so that the coverage is better.

OK I did some tests and corrected the uploader part in process_clicked.

fernandezc commented on spectrochempy/widgets/baselinecorrector.py:
```diff
@@ -0,0 +1,311 @@
+    The `BaselineCorrector()` widget can be used in jupyter notebooks.
+    - The GUI buttons are as follows:
+        - `upload`: upload files (disabled if a NDDataset is passed as parameter).
+          Uploading file will not trigger the reading
```
I am wondering what is the reason of not triggering the processing

atravert replied:
because I was lazy :-). Now it also triggers reading/processing (see last sereis of commits [06da720](https://github.com/spectrochempy/spectrochempy/pull/397/commits/06da720a32e3aba7ce13399c897f11583fef6d7c) - [a3b873c](https://github.com/spectrochempy/spectrochempy/pull/397/commits/a3b873c2744f77fbe39841b69e4ab7240751d8da)). 
With this change, it still works in jupyter notebook, but not in jupyter lab (plots are visible in the log, though ;   so I don't know how to debug this ).

fernandezc replied:
I will try to look at this this afternoon.

fernandezc commented on spectrochempy/widgets/baselinecorrector.py:
```diff
@@ -0,0 +1,311 @@
+        - `save as`: save the baseline corrected NDDataset
+
+    - The `x slice` and `y slice` textboxes can be used to slice the initial
+      dataset with the usual `[start:stop:stride]`
```
step instead of stride

atravert replied:
done

fernandezc commented:
For the moment, I don't think I'll add anything, but of course if my modifications are questionable, don't hesitate. I didn't touch the doc and the comments which are not necessarily up to date, especially on the slicing.

fernandezc commented:
It's very strange, if you use the uploader, it doesn't work (no display) in Jupyter Lab. No problem if the file is passed in the `init`.
  If you upload the non-existent file using for example:
```
if self._x is None:
    self._X = read()
```
in this case the usual dialog is opened and the rest is executed normally. The problem is related to FileUpLoad. I don't know why, I looked everywhere  but no error message. The only one is in the Jupiter LAb log in the console, a series of : 
```
[IPKernelApp] WARNING | No such comm: 8fe14cadd2574a44a4dff6b79beab836
```
But I didn't understand how this can be fixed. 

So either we give up the idea of being able to do an upload in jupyter lab (it's a shame I prefer this environment to notebook). Or we automatically execute a `read` as soon as the init doesn't find an input file, as explained above. 

Otherwise, the tests should be complete because there is a little bit of new code that is not covered: I was at 99%! we are back to 92%. It's not bad, but I saw that especially in what you added on the ranges, we can certainly do better.

fernandezc commented:
We could also write slightly different code for the Jupyter lab.  But I haven't figured out how to determine where we are running (Lab or Notebook) from python. The only thing I do in the spectroscopy init is determine if we are in IPython or not. I found a few solutions on the web, but they seem to be very system dependent.

fernandezc commented:
Last thing, it seems to me that in the notebook, it would be necessary to disable the uplaod button after loading, because there are errors if you try a second time. Or you have to manage it in the code.

atravert commented:
Hi,
> or we automatically execute a `read` as soon as the init doesn't find an input file, as explained above.

Yes I agree, it is actually what I did in the very first versions of the widget :-), I have replaced it by a FileLoader because the with scp.read(), on my system, the pop up window does not aloways appear on the current screen (when I use multi-screen, I tghink it firste appear on the default screen (1), not necessarily the one where the browser window. I tried long tilme ago to force the focus of filedialog windows, but I remmber much :-). Anyway I agrre with this proposal, abd it will give a more consistent  user experience with same filedialogues for load and save :-) 

> it would be necessary to disable the uplaod button after loading, because there are errors if you try a second time.
it does not cause error in may case (the old files are replaced by neaw files), but yes, i will do it,  and it will be more clear for the user. 

> Otherwise, the tests should be complete because there is a little bit of new code that is not covered: I was at 99%! we are back to 92%.

YEs ! I will do that too :-)

Best,
Arnaud

atravert commented:
PS:
> I tried long time ago to force the focus of filedialog windows, but I remember much :-).

because the filedialog does not in,itially appean on the foreground, which is painfuil when a lot of apps are running... but as this is not sepecific to the widget, it can be solved later :-)

atravert commented:
done 99% coverage + replacement of uploader by read(). As far as I remember it was OK with jupyter lab at the begining, but not anymore... I give up with jupyer lab compatibility for the moment to focus on other stuff !

fernandezc commented:
>  I give up with jupyter lab compatibility for the moment to focus on other stuff !

Yes, I understand. I spent a little too much time on that too.
I'm going to look anyway if using another library like Plotly or bqplot would not allow a better compatibility.

fernandezc commented:
To reassure you about Jupyter Lab, your last modifications don't work for me anymore in Jupyter Lab: the simple fact to have a button to click before the display seems to make the display impossible. So it's not FileUpLoad that's the problem: Indeed if you do as I suggested below, it works: 

Line 171:
```
        self._input = widgets.HBox(children=[io, controls])
        self._output = widgets.Output()
        display(self._input)

        if self._X is None:
            self._X = read()

        if self._X is not None:
            self.process_clicked()
        else:
            self.corrected = NDDataset()
```

Since I can't be satisfied with this situation (I'm very stubborn!) I'm going to dig a little deeper.

fernandezc commented on docs/userguide/processing/baseline.py:
```diff
@@ -276,6 +588,34 @@
+# - The `x slice` anbd `y slice` textboxes can be used to slice the initial dataset with the usual `[start:stop:step]`
+# format. In the `x` dimension, coordinates or indexes can be used (e.g. `[3000.0:2000.0:1]` or `[0:100:1]` are valid
+# entries). In the `y` dimension only indexes can be used (e.g. `[0:10:1]`). Note also that currently none of the
+# `start`, `stop`, `step` parameters can be omitted, e.g. `[3000.0:2000.0]` or `[:,:]` are not valid entries.
```
This is not True anymore: you can enter as in numpy:
`[start:stop:step]`
`[:stop:step]`
`[::step]`
`[start::step]`
etc...

and `start:` etc.... without the bracket

fernandezc replied:
and this for x and y, and also you can use index or location for both.

atravert commented:
ahaha, this is a quality in the present case :-)

Maybe in the case where self_X is None we could initialize output with 
two empty plots ?


Le 07/03/2022 à 15:47, Christian Fernandez a écrit :

> To reassure you about Jupyter Lab, your last modifications don't work 
> for me anymore in Jupyter Lab: the simple fact to have a button to 
> click before the display seems to make the display impossible. So it's 
> not FileUpLoad that's the problem: Indeed if you do as I suggested 
> below, it works:
>
> Line 171:
>
> |self._input = widgets.HBox(children=[io, controls]) self._output = 
> widgets.Output() display(self._input) if self._X is None: self._X = 
> read() if self._X is not None: self.process_clicked() else: 
> self.corrected = NDDataset() |
>
> Since I can't be satisfied with this situation (I'm very stubborn!) 
> I'm going to dig a little deeper.
>
> —
> Reply to this email directly, view it on GitHub 
> <https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fspectrochempy%2Fspectrochempy%2Fpull%2F397%23issuecomment-1060766671&data=04%7C01%7Carnaud.travert%40ensicaen.fr%7Cd935d26cc7ac4924db6208da00496d7a%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C637822612660866761%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=jTj7i7QZBXiM17wTbDHajg8oXInczVJbZPVtl98grJ8%3D&reserved=0>, 
> or unsubscribe 
> <https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FACNBNCXVTCJGLUAWYPYR4VTU6YJIVANCNFSM5PKE7BBA&data=04%7C01%7Carnaud.travert%40ensicaen.fr%7Cd935d26cc7ac4924db6208da00496d7a%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C637822612660866761%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=g9GhzxrSNaEbXuon3aMyUZVjLtarxNYK8FIBxlDfxxk%3D&reserved=0>.
> Triage notifications on the go with GitHub Mobile for iOS 
> <https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fapps.apple.com%2Fapp%2Fapple-store%2Fid1477376905%3Fct%3Dnotification-email%26mt%3D8%26pt%3D524675&data=04%7C01%7Carnaud.travert%40ensicaen.fr%7Cd935d26cc7ac4924db6208da00496d7a%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C637822612660866761%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=WpzzZlb%2F0ovQVrvHATO6xfl8xJqSCCSn9kgoEDzc8%2FY%3D&reserved=0> 
> or Android 
> <https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fplay.google.com%2Fstore%2Fapps%2Fdetails%3Fid%3Dcom.github.android%26referrer%3Dutm_campaign%253Dnotification-email%2526utm_medium%253Demail%2526utm_source%253Dgithub&data=04%7C01%7Carnaud.travert%40ensicaen.fr%7Cd935d26cc7ac4924db6208da00496d7a%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C637822612660866761%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=25gXbdqPejw91bSRwUeqhN4GJiSbRFihyZrrd%2Bg6vuE%3D&reserved=0>. 
>
> You are receiving this because you modified the open/close 
> state.Message ID: 
> ***@***.***>
>

fernandezc commented:
Well, I think I've got it! In order to have the figure displayed explicitly in the `self._output` object you have to use the `show` method of self._fig. If you don't do it, the figure is displayed outside `self._output` (you can see that by putting a frame around the object: `layout={'border': '.1px solid black'}`), and then in jupyter lab it doesn't work anymore.
